smqtk_descriptors/impls/descriptor_generator/prepopulated.py

```diff
@@ -0,0 +1,124 @@
+# copy of pytorch
+import abc
+import copy
+import itertools
+import logging
+from typing import (
+    Any, Callable, Dict, Iterable, Iterator, Literal, Mapping,
+    Optional, Sequence, Set, Type, TypeVar, Union
+)
+
+import numpy as np
+
+from smqtk_dataprovider import DataElement
+from smqtk_descriptors import DescriptorGenerator
+from smqtk_image_io import ImageReader
+from smqtk_core.configuration import (
+    from_config_dict,
+    make_default_config,
+    to_config_dict,
+)
+from smqtk_descriptors.utils.parallel import parallel_map
+
+LOG = logging.getLogger(__name__)
+
+__all__ = [
+    "PrePopulatedDescriptorGenerator"
+
+]
+T = TypeVar("T", bound="PrePopulatedDescriptorGenerator")
+
+
+
+
+class PrePopulatedDescriptorGenerator (DescriptorGenerator):
+
+
+    @classmethod
+    def is_usable(cls) -> bool:
+        return True
+
+    @classmethod
+    def get_default_config(cls) -> Dict[str, Any]:
+        c = super().get_default_config()
+        return c
+
+    @classmethod
+    def from_config(
+        cls: Type[T],
+        config_dict: Dict,
+        merge_default: bool = True
+    ) -> T:
+        # Copy config to prevent input modification
+        config_dict = copy.deepcopy(config_dict)
+        return super().from_config(config_dict, merge_default)
+
+    def __init__(
+        self,
+    ):
+        super().__init__()
+
+    @abc.abstractmethod # might be internal and never called
+    def _load_module(self) -> "torch.nn.Module":
+        """
+        :return: Load and return the module.
+        """
+
+    def __getstate__(self) -> Dict[str, Any]:
+        return self.get_config()
+
+    def __setstate__(self, state: Mapping[str, Any]) -> None:
+        # This ``__dict__.update`` works because configuration parameters
+        # exactly match up with instance attributes currently.
+        self.__dict__.update(state)
+
+
+
+    def valid_content_types(self) -> Set:
+        return set()
+
+    def is_valid_element(self, data_element: DataElement) -> bool:
+        # Check element validity though the ImageReader algorithm instance
+        return False
```

Comment:
Between this and the `generate_arrays_*` methods all raising exceptions, is the intent of this implementation to just raise errors whenever there is an attempt to use it? If so, that is something that should go into the doc-string for the class.

This is on the boarder of bike-shedding, but in such a case, I almost feel like the class name could be something more appropriate for its generic behavior around intentionally not running, e.g. "ErrorGenerator" or "ErrorDescriptorGenerator".

Reply:
The name is indicating that if you are using this in the config, then you expect that the descriptor cache is prepopulated. Thus if you are trying to generate a descriptor at any point, something is going wrong. We should consolidate this into the class-level docstring. It might also make sense to change the `raise NotImplementedError`s to `raise AssertionError`.